### PR TITLE
Add URL matcher functionality for domain-based application selection

### LINKED ISSUE #75: Domain Specific Applications Possible?
I would like Junction to provide [FreeTube](https://github.com/FreeTubeApp/FreeTube) as an app to open YouTube urls and only YouTube urls. Would there be a way to get Junction to recognize FreeTube as such? Whether that be through something is Junction or in the system MIME database. Currently it registers itself as a handler of the MIME type `x-scheme-handler/freetube`, idk if that helps.

Thank you for your time.

### PR DESCRIPTION
**Pull request by lwille**
## Summary

This PR implements URL matcher functionality to automatically route URLs to specific applications based on user-defined regex patterns, addressing #75.

### Key Features

- **URL Pattern Matching**: Users can define regex patterns to match URLs
- **Automatic Application Selection**: Matching URLs automatically open in the configured application
- **Preferences Integration**: New Preferences button in welcome screen for managing URL matchers
- **GSettings Storage**: URL matchers are persisted using GSettings with proper serialization
- **Libadwaita UI**: Modern dialog components following GNOME design guidelines

### Implementation Details

- Added `url-matchers` GSettings key for storing JSON-serialized patterns
- Implemented `getUrlMatchers()`, `addUrlMatcher()`, `removeUrlMatcher()` functions in `util.js`
- Added `findMatchingApp()` for automatic URL-to-application matching
- Enhanced `window.js` to check for URL matches before displaying application chooser
- Refactored preferences action to use application-level `Gio.SimpleAction`
- Replaced deprecated `GtkShortcutsWindow` with modern `Adw.Dialog`

### Use Cases

- Open GitHub URLs in GitHub Desktop: `https://github\.com/.*`
- Route YouTube to media player: `https://youtube\.com/.*`
- Direct documentation to specific browser: `https://docs\..*`
- Route work tools to work browser: `.*\.company\.com/.*`

### Test Coverage

- Comprehensive URL pattern matching tests
- JSON serialization/deserialization validation
- Edge case handling for invalid patterns

## Test Plan

- [ ] Verify Preferences button appears in welcome screen
- [ ] Test URL matcher creation and deletion via preferences
- [ ] Confirm automatic URL routing works for HTTP/HTTPS schemes
- [ ] Validate pattern validation and error handling
- [ ] Test GSettings persistence across app restarts

Fixes #75

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: src/preferences.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    });
+
+    row.add_suffix(deleteButton);
+
+    return row;
+  }
+
+  function showErrorDialog(message) {
+    const errorDialog = new Adw.AlertDialog({
+      heading: "Error",
```
Comment (sonnyp) [inline src/preferences.js]:
```suggestion
      heading: _("Error"),
```

File: src/preferences.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      body: message,
+    });
```
Comment (sonnyp) [inline src/preferences.js]:
```suggestion
      "close-response": "ok",
    });
```

File: src/preferences.js [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    errorDialog.add_response("ok", "OK");
+    errorDialog.present(preferences_window);
+  }
+
+  preferences_window.present(null);
```
Comment (sonnyp) [inline src/preferences.js]:
```suggestion
  preferences_window.present(application.active_window);
```

File: src/preferences.blp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+          margin-bottom: 24;
+        }
+      }
+    }
+  }
+}
+
+Adw.AlertDialog add_matcher_dialog {
+  heading: _("Add URL Matcher");
+  body: _("Enter a regular expression pattern to match URLs and select the application to open them with.");
```
Comment (sonnyp) [inline src/preferences.blp]:
```suggestion
```
UI should be clear enough that this is not necessary

File: src/preferences.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  add_matcher_dialog.add_response("add", "Add");
+  add_matcher_dialog.set_response_appearance("add", Adw.ResponseAppearance.SUGGESTED);
+
+  // Populate browser applications
+  populateAppDropdown();
+
+  // Load existing matchers
+  loadMatchers();
+
+  // Add button handler
```
Comment (sonnyp) [inline src/preferences.js]:
```suggestion
```

File: src/preferences.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  add_button.connect("clicked", () => {
+    pattern_entry.set_text("");
+    app_dropdown.set_selected(0);
+    add_matcher_dialog.present(preferences_window);
+  });
+
+  // Dialog response handler
```
Comment (sonnyp) [inline src/preferences.js]:
```suggestion
```

File: src/preferences.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+  add_matcher_dialog.connect("response", (dialog, response) => {
+    if (response === "add") {
```
Comment (sonnyp) [inline src/preferences.js]: please prefer eary return

ie

```
if (response !== "add") {
  return dialog.close()
}
...
```

File: src/PreferencesDialog.blp [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        subtitle: _("Create a new rule to automatically open matching URLs");
+        
+        [suffix]
+        Button add_button {
+          icon-name: "list-add-symbolic";
+          tooltip-text: _("Add new URL matcher");
+          valign: center;
+          styles ["flat"]
+        }
+      }
```
Comment (sonnyp) [inline src/PreferencesDialog.blp]: This shouldn't be a row

Instead implement this a pill button above the ListBox https://developer.gnome.org/hig/patterns/containers/boxed-lists.html#adding-buttons

File: src/PreferencesDialog.blp [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+      ListBox matchers_list {
+        styles ["boxed-list"]
+        selection-mode: none;
+        
+        Label empty_label {
+          label: _("No URL matchers configured.\nClick the + button to add one.");
+          styles ["dim-label"]
+          margin-top: 24;
+          margin-bottom: 24;
+        }
```
Comment (sonnyp) [inline src/PreferencesDialog.blp]: this is not visible anywhere

File: src/PreferencesDialog.js [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+  // Dialog response handler
+  add_matcher_dialog.connect("response", (dialog, response) => {
+    if (response === "add") {
+      const pattern = pattern_entry.get_text();
+      const selectedIndex = app_dropdown.get_selected();
+
+      if (pattern && selectedIndex !== Gtk.INVALID_LIST_POSITION) {
+        const appString = app_list_model.get_string(selectedIndex);
+        const [appId] = appString.split(" - ");
```
Comment (sonnyp) [inline src/PreferencesDialog.js]: too big of a hack

File: src/PreferencesDialog.blp [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+      }
+    }
+  }
+}
+
+Adw.AlertDialog add_matcher_dialog {
+  heading: _("Add URL Matcher");
+  body: _("Enter a regular expression pattern to match URLs and select the application to open them with.");
```
Comment (sonnyp) [inline src/PreferencesDialog.blp]: UI should be self explanatory


```suggestion
```
